web/source/keymanweb.js

```diff
@@ -3312,51 +3342,52 @@ if(!window['keyman']['initialized']) {
     * 
     * @param      {number|boolean}  bBack     Direction to move (0 or 1)
     */
-    keymanweb.moveToNext=function(bBack)
-    {
-      var i,t=keymanweb.sortedInputs,
-        activeBase=keymanweb._ActiveElement;
+    keymanweb.moveToNext=function(bBack) {
+      var i,t=keymanweb.sortedInputs, activeBase=keymanweb._ActiveElement;
       
-      if(t.length == 0) return;
+      if(t.length == 0) {
+        return;
+      }
 
       // For touchable devices, get the base element of the DIV
-      if(device.touchable) activeBase=activeBase.base;
+      if(device.touchable) {
+        activeBase=activeBase.base;
+      }
 
       // Identify the active element in the list of inputs ordered by position
-      for(i=0; i<t.length; i++)
-      {          
+      for(i=0; i<t.length; i++) {          
         if(t[i] == activeBase) break;
       }   
 
       // Find the next (or previous) element in the list
-      if(bBack) i=i-1; else i=i+1; 
-      if(i >= t.length) i=i-t.length;
-      if(i < 0) i=i+t.length;
+      if(bBack) {
+        i=i-1;
+      } else {
+        i=i+1;
+      }
+      if(i >= t.length) {
+        i=i-t.length;
+      }
+      if(i < 0) {
+        i=i+t.length;
+      }
 
       // Move to the selected element
-      if(device.touchable)
-      {                     
+      if(device.touchable) {                     
         // Set focusing flag to prevent OSK disappearing 
         keymanweb.focusing=true;
         var target=t[i]['kmw_ip'];
 
         // Focus if next element is non-mapped
-        if(typeof(target) == 'undefined')
-        {
+        if(typeof(target) == 'undefined') {
           t[i].focus();
-        }
-        
-        // Or reposition the caret on the input DIV if mapped
-        else  
-        {
+        } else { // Or reposition the caret on the input DIV if mapped
           keymanweb._ActiveElement=keymanweb._LastActiveElement=target;    
           keymanweb.setTextCaret(target,10000);                            
-          keymanweb.scrollInput(target);                                   
+          keymanweb.scrollInput(target);   // mousedown check
+          target.focus();
```

Comment:
This line here is new behaviour. I don't quite understand why it is needed?

Reply:
Previously, switching between touch-based elements always happened with the OSK, which didn't need to intercept keystrokes.  It turns out that to properly support keystrokes, we needed `focus` to be called here to ensure things don't break upon use of the ENTER key within single-line controls, or if using the OSK to swap and then typing.